oauth/scope.js

```diff
@@ -0,0 +1,514 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+const { URL } = require('url');
+
+// These character ranges are from the OAuth RFC,
+// https://tools.ietf.org/html/rfc6749#section-3.3
+const VALID_SCOPE_VALUE = /^[\x21\x23-\x5B\x5D-\x7E]+$/;
+
+const VALID_SHORT_NAME_VALUE = /^[a-zA-Z0-9_]+$/;
+const VALID_FRAGMENT_VALUE = /^#[a-zA-Z0-9_]+$/;
+
+
+/**
+ * A `Scope` is a set of validated scope string values, against which
+ * we can perform various membership checks and set-like operations
+ * such as:
+ *
+ *  - Checking whether one set of scope values wholly or partly
+ *    contains another set.
+ *
+ *  - Filtering out scope values that are (or are not) implied
+ *    by another set.
+ *
+ *  - Aggregating values into a single set.
+ *
+ * We expect `Scope` objects will be used to process user-provided
+ * data, so it's important that they provide efficient operations
+ * that do not open potential DoS vectors.  A naive item-by-item
+ * comparison of two sets can easily be accidentally quadratic.
+ *
+ * The key insight required for achieving this is that, for given
+ * a scope string `v`, we can enumerate the set of all possible values
+ * that imply `v`, and the size of this set of linear in the length
+ * of `v`.  These values are called the "implicants" of `v` and
+ * by pre-calculting and storing them, we are able to perform many
+ * `Scope` operations using simple set lookups.
+ *
+ * To that end, `Scope` instances store their data internally as a
+ * pair of cross-referenced maps, `this._values` and `this._implicants`,
+ * both of which are keyed by scope string values.
+ *
+ * The `_values` map has as its keys the actual scope values contained
```

Comment:
Ahha, here is the documentation on how _values and _implicants are used in tandem, I read and then immediately forgot this the first time.